exonum/src/events/noise/mod.rs

```diff
@@ -112,3 +114,174 @@ fn write_handshake_msg(
     let res = noise.write_handshake_msg();
     Box::new(done(res.map_err(|e| e.into())))
 }
+
+pub fn read_handshake_msg(
+    input: &[u8],
+    noise: &mut NoiseWrapper,
+) -> Box<Future<Item = (usize, Vec<u8>), Error = io::Error>> {
+    let res = noise.read_handshake_msg(input);
+    Box::new(done(res.map_err(|e| e.into())))
+}
+
+#[cfg(test)]
+mod tests {
+    use tokio_core::reactor::Core;
+    use tokio_core::net::{TcpListener, TcpStream};
+    use tokio_io::AsyncRead;
+    use futures::{done, Future, Stream};
+
+    use std::net::SocketAddr;
+    use std::thread;
+    use std::io;
+
+    use events::noise::wrapper::{NoiseWrapper, NOISE_MAX_MESSAGE_LENGTH,
+                                 NOISE_MIN_HANDSHAKE_MESSAGE_LENGTH};
+    use events::noise::{read, read_handshake_msg, write, HandshakeParams, HandshakeResult,
+                        NoiseHandshake};
+    use events::error::{into_other, log_error};
+    use events::codec::MessagesCodec;
+    use crypto::{gen_keypair_from_seed, Seed};
+    use std::time::Duration;
+
+    #[derive(Debug, PartialEq, Copy, Clone)]
+    pub enum HandshakeStep {
+        Default,
+        One(u8, usize),
+        Two(u8, usize),
+    }
+
+    const EMPTY_MESSAGE: usize = 0;
+    const SMALL_MESSAGE: usize = NOISE_MIN_HANDSHAKE_MESSAGE_LENGTH - 1;
+    const BIG_MESSAGE: usize = NOISE_MAX_MESSAGE_LENGTH + 1;
+
+    impl Default for HandshakeParams {
+        fn default() -> Self {
+            let (public_key, secret_key) = gen_keypair_from_seed(&Seed::new([0; 32]));
+            HandshakeParams {
+                max_message_len: 1024,
+                public_key,
+                secret_key,
+            }
+        }
+    }
+
+    #[test]
+    fn test_noise_normal_handshake() {
+        let addr: SocketAddr = "127.0.0.1:45001".parse().unwrap();
+        let addr2 = addr.clone();
+
+        thread::spawn(move || {
+            run_handshake_listener(&addr2).unwrap();
+        });
+
+        // Wait until listener is initialized.
+        thread::sleep(Duration::from_millis(500));
+
+        let res = send_handshake(&addr, HandshakeStep::Default);
+        assert!(res.is_ok());
+    }
+
+    #[test]
+    fn test_noise_handshake_errors() {
+        let addr: SocketAddr = "127.0.0.1:45002".parse().unwrap();
+        let addr2 = addr.clone();
+
+        // Wait until listener is initialized.
+        thread::sleep(Duration::from_millis(500));
+
+        thread::spawn(move || run_handshake_listener(&addr2));
+
+        let res = send_handshake(&addr, HandshakeStep::One(1, EMPTY_MESSAGE));
+        assert!(res.is_err());
```

Comment:
The logic of send_handshake is too complicated, is it possible to check for a concrete error?

And this would be good if you will test `max_message_len` too. 
And probably `HandshakeStep::One` \  `HandshakeStep::Two` with valid message size should behave somehow too.